cpp/src/io/json/json_tree.cu

```diff
@@ -254,12 +308,435 @@ tree_meta_t get_tree_representation(device_span<PdaTokenT const> tokens,
+#ifdef NJP_DEBUG_PRINT
+  print_vec(level_boundaries, "level_boundaries");
+  std::cout << "num_levels: " << num_levels << std::endl;
+#endif
+
+#ifdef NJP_DEBUG_PRINT
+  auto print_level_data = [stream](auto stage_text,
+                                   auto level,
+                                   auto start,
+                                   auto end,
+                                   auto const& nodeII,
+                                   auto const& parent_node_idx,
+                                   auto const& parent_col_id,
+                                   auto const& node_type,
+                                   auto const& levels,
+                                   auto const& col_id) {
+    std::cout << level << stage_text << "\n";
+    auto cls = std::array<const char*, 6>{"S", "L", "F", "R", "V", "E"};
+    for (auto n = start; n < end; n++)
+      printf("%3d ", nodeII.element(n, stream));
+    printf(" nodeII-%ld\n", level);
+    for (auto n = start; n < end; n++)
+      printf("%3d ", parent_node_idx.element(n, stream));
+    printf(" parent_node_idx-%ld\n", level);
+    for (auto n = start; n < end; n++)
+      printf("%3d ", parent_col_id.element(n, stream));
+    printf(" parent_col_id-%ld\n", level);
+    for (auto n = start; n < end; n++) {
+      auto nt = node_type.element(n, stream);
+      printf("%3s ",
+             nt >= NUM_NODE_CLASSES ? std::to_string(nt - NUM_NODE_CLASSES).c_str() : cls[nt]);
+    }
+    printf(" node_type-%ld\n", level);
+    for (auto n = start; n < end; n++)
+      printf("%3d ", levels.element(n, stream));
+    printf(" levels-%ld\n", level);
+    for (auto n = start; n < end; n++)
+      printf("%3d ", col_id.element(n, stream));
+    printf(" col_id-%ld\n", level);
+  };
+#define PRINT_LEVEL_DATA(level, text)           \
+  print_level_data(text,                        \
+                   level,                       \
+                   level_boundaries[level - 1], \
+                   level_boundaries[level],     \
+                   scatter_indices,             \
+                   parent_indices,              \
+                   parent_col_id,               \
+                   node_type,                   \
+                   d_tree.node_levels,          \
+                   col_id);
+#else
+#define PRINT_LEVEL_DATA(level, text) ;
+#endif
+
+  // 4. Propagate parent node ids for each level.
+  // For each level,
+  //     a. gather col_id from previous level results. input=col_id, gather_map is parent_indices.
+  //     b. stable sort by {parent_col_id, node_type}
+  //     c. scan sum of unique {parent_col_id, node_type}
+  //     d. scatter the col_id back to stable node_level order (using scatter_indices)
+  nvtxRangePushA("pre-level");
+  rmm::device_uvector<NodeIndexT> col_id(num_nodes, stream, mr);
+  rmm::device_uvector<NodeIndexT> parent_col_id(num_nodes, stream);
+  thrust::uninitialized_fill(rmm::exec_policy(stream),
```

Comment:
Not familiar enough with thrust to answer: why uninitialized_fill and not fill? Does it matter whether a primitive type value was previously uninitialized? I'd expect the generated code to be identical

Reply:
It does not matter here. will replace with fill.
uninitialized_fill uses constructor on the memory. fill uses assignment operator.